### DISCUSSION TITLE
Performance Gotenberg VS Libreoffice

### DISCUSSION THREAD

**Discussion by ytilotti** [gotenberg/gotenberg#526] [Q&A]
Hi team. 

I set up Gotenberg (7.6.1) on my server as a Docker daemon and installed Libreoffice directly on my server in parallel.
I have trouble understanding the performance differences. For 4 files conversion, I am at 12 seconds on Gotenberg and it is almost instantaneous on Libreoffice.

Is there configuration specific on Docker stack? Have you more information on this?

The installation of libreoffice:
```
mkdir -p /usr/share/man/man1
apt-get -y install default-jdk-headless libreoffice-core libreoffice-writer libreoffice-calc libreoffice-impress
mkdir -p /var/www/.cache/dconf
mkdir -p /var/www/.config/libreoffice
chmod -R ugo+rwx /var/www/.cache
chmod -R ugo+rwx /var/www/.config
```

**Comment by gulien** [comment]
Hello @ytilotti,

In stateless mode, Gotenberg starts a dedicated LibreOffice instance per conversion, while in stateful mode, Gotenberg starts one long-running instance that can handle one and only one conversion at a time.

In other words, stateless mode works better for bulk conversions, while stateful mode is quicker for single conversion.

How did you convert with your local LibreOffice?

**Reply by ytilotti** [reply]
I'm on PHP env on this feature.

With local LibreOffice, I use the SDK [mnvx/lowrapper](https://github.com/mnvx/lowrapper). The SDK only use the executable Linux via `exec()`.
With Gotenberg, I use the SDK in PHP.

Send the content of documents to the Docker, can maybe slower the conversion? But we are in local environnement, no impact.
I don't understand this big difference in execution.

**Reply by gulien** [reply]
I’m using `unoconv` as a uno API wrapper, and it handles a bit more than the default LibreOffice executable (for instance, PDF archive). I’m planning to move out of `unoconv`, but for now I don’t have enough time available.

Regarding the installation, it’s more or less the same but I also install the whole package, and I do not have these lines:

```
mkdir -p /var/www/.cache/dconf
mkdir -p /var/www/.config/libreoffice
chmod -R ugo+rwx /var/www/.cache
chmod -R ugo+rwx /var/www/.config
```

I’m curious if they do make some differences (note that they should use another path for my case, as the working directory is different in the Gotenberg image.

If you have some time to test things out, I’d be very curious about the results ^^

**Reply by ytilotti** [reply]
There is an error on Dockerfile or it's me?
`mkdir -p /usr/share/man/man1mkdir -p /usr/share/man/man1 &&\`

**Reply by ytilotti** [reply]
I tested via a build, change nothing.

```
FROM gotenberg/gotenberg:7

USER root

RUN mkdir -p /var/www/.cache/dconf
RUN mkdir -p /var/www/.config/libreoffice
RUN chmod -R ugo+rwx /var/www/.cache
RUN chmod -R ugo+rwx /var/www/.config

USER gotenberg
```

**Reply by gulien** [reply]
It is legit actually: https://github.com/gotenberg/gotenberg/blob/b73207f29abf2b573d21edcdb32e82f15070dfbe/build/Dockerfile#L140

`/var/www` seems actually correlated with the working directory `/var/www/html`. So in our case `/home` may be the correct path as Gotenberg working directory is `/home/gotenberg` (but that’s just my guess).

**Reply by ytilotti** [reply]
It's a typo in this [comment](https://github.com/nextcloud/docker/issues/380#issuecomment-409593925), only this is necessary:
`mkdir -p /usr/share/man/man1`

The command `mkdir -p /usr/share/man/man1mkdir -p /usr/share/man/man1` create 2 dir: `man1mkdir` & `man1` (more `-p`  in root). It's a luck that it works!

**Reply by ytilotti** [reply]
> `/var/www` seems actually correlated with the working directory `/var/www/html`. So in our case `/home` may be the correct path as Gotenberg working directory is `/home/gotenberg` (but that’s just my guess).

You have right sorry. But change nothing.

```
FROM gotenberg/gotenberg:7

USER root

RUN mkdir -p /home/gotenberg/.cache/dconf
RUN mkdir -p /home/gotenberg/.config/libreoffice
RUN chmod -R ugo+rwx /home/gotenberg/.cache
RUN chmod -R ugo+rwx /home/gotenberg/.config

USER gotenberg
```

**Reply by gulien** [reply]
Good catch with the typo, you caught my brainless c/c lol

If that’s change nothing, I guess it can be narrowed down to:

1. `unoconv` uno API is not as fast as direct command via the executable.
2. And/or Gotenberg does more processing (i.e. download the file, validate inputs, call uno, send back the resulting PDF).

**Reply by ytilotti** [reply]
Hi @gulien.

Can you confirm me, but the content of the files is send to the docker stateless Gotenberg? It's not the path because the docker can't acces to the files.

Edit : The files are send via form with multipart/form-data. This confirm my question.

**Reply by gulien** [reply]
Yep Gotenberg handles form data, the paths in the documentation examples are just how curl handle the creation of a `multipart/form-data` with files content ^^

**Reply by ytilotti** [reply]
Have you try/test a solution with volume? Just a question.

**Reply by gulien** [reply]
Not really, because, from an infrastructure perspective, it is easier to have a brick (i.e., Gotenberg) that does not depend on another external brick (volume, file system, database, etc.).

**Reply by ytilotti** [reply]
It was a reflection on a possible test.

What is your execution time to convert [libreoffice/sample1.docx](https://github.com/gotenberg/gotenberg/blob/main/test/testdata/libreoffice/sample1.docx) in pdf? 4s here.

**Reply by gulien** [reply]
You may test it on the demo instance: https://demo.gotenberg.dev/forms/libreoffice/convert 👍

**Reply by ytilotti** [reply]
Less than 1 second with same file! Don't understand. There is a configuration specific on the docker? Last version on the demo?

**Reply by gulien** [reply]
No, the demo instance is actually the standard Docker image (plus a rate limiting middleware). It’s also using the stateful mode, which is not really stable but offers more performance for unitary conversion, as you know.

**Reply by ytilotti** [reply]
Ok i found the problem. I thought it was optimised in that way, but not.
If you do this:

```
curl \
--request POST 'http://localhost:3000/forms/libreoffice/convert' \
--form 'files=@"lorem-ipsum.pdf"' \
--form 'files=@"docx.docx"' \
--form 'files=@"lorem-ipsum_1.pdf"' \
--form 'files=@"lorem-ipsum_2.pdf"' \
--form 'files=@"lorem-ipsum_3.pdf"' \
--form 'merge="true"' \
-o my.pdf
```
This previous example is very slow compared to this next couple.
```
curl \
--request POST 'http://localhost:3000/forms/libreoffice/convert' \
--form 'files=@"docx.docx"' \
-o docx.pdf

curl \
--request POST 'http://localhost:3000/forms/pdfengines/merge' \
--form 'files=@"lorem-ipsum.pdf"' \
--form 'files=@"docx.pdf"' \
--form 'files=@"lorem-ipsum_1.pdf"' \
--form 'files=@"lorem-ipsum_2.pdf"' \
--form 'files=@"lorem-ipsum_3.pdf"' \
-o my.pdf
```

Can be interesting to do an optimisation in this way.

**Reply by gulien** [reply]
Oh yeah totally forgot to mention that there is no first start (in stateful mode), meaning the first request will start the LibreOffice instance and it will run until the threshold is reached.
